extensions/amp-ad-exit/0.1/amp-ad-exit.js

```diff
@@ -237,6 +261,101 @@ export class AmpAdExit extends AMP.BaseElement {
+        `${TAG}: No friendly parent amp-ad element was found for amp-ad-exit.`);
+
+    this.init3pResponseListener_();
+  }
+
+  /**
+   * Gets the resource ID of the amp-ad element containing this amp-ad-exit tag.
+   * This is a pass-through for the version in service.js, solely because
+   * the one in service.js isn't stubbable for testing, since only object
+   * methods are stubbable.
+   * @returns {?string}
+   * @private
+   * @VisibleForTesting
+   */
+  getAmpAdResourceId_() {
+    return getAmpAdResourceId(this.element, this.win.top);
+  }
+
+  /** @override */
+  resumeCallback() {
+    this.init3pResponseListener_();
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    if (this.unlisten_) {
+      this.unlisten_();
+      this.unlisten_ = null;
+    }
+    return super.unlayoutCallback();
+  }
+
+  /**
+   * amp-analytics will create an iframe for vendors in
+   * extensions/amp-analytics/0.1/vendors.js who have transport/iframe defined.
+   * This is limited to MRC-accreddited vendors. The frame is removed in
+   * amp-analytics, and the listener is destroyed here, if the user
+   * navigates/swipes away from the page. Both are recreated if the user
+   * navigates back to the page.
+   * @private
+   */
+  init3pResponseListener_() {
+    this.unlisten_ = this.unlisten_ || listen(this.getAmpDoc().win, 'message',
+        event => {
+          const responseMsg = deserializeMessage(getData(event));
+          if (!responseMsg ||
+              responseMsg['type'] != MessageType.IFRAME_TRANSPORT_RESPONSE) {
+            return;
+          }
+          this.assertValidResponseMessage_(responseMsg, event.origin);
+          this.vendorResponses_[responseMsg['vendor']] = responseMsg['message'];
+        });
+  }
+
+  /**
+   *
+   * @param {JsonObject} responseMessage The response object to
+   *     validate.
+   * @param {string} responseOrigin The origin of the message, which should
+   *     match the expected origin of messages from responseMessage['vendor']
+   * @private
+   */
+  assertValidResponseMessage_(responseMessage, responseOrigin) {
+    user().assert(responseMessage['message'],
+        'Received empty response from 3p analytics frame');
+    user().assert(
+        responseMessage['creativeId'] == this.ampAdResourceId_,
+        'Received malformed message from 3p analytics frame: ' +
+        'creativeId missing or invalid');
+    user().assert(responseMessage['vendor'],
+        'Received malformed message from 3p analytics frame: ' +
+        'vendor missing');
+    this.assertOriginMatchesVendor_(responseOrigin, responseMessage['vendor']);
+  }
+
+  /**
+   * Ensures that a given origin matches that of an existing vendor's
+   * transport/iframe URL
+   * @param {string} origin The origin to verify
+   * @param {string} vendor The vendor whose origin to check against
+   * @private
+   */
+  assertOriginMatchesVendor_(origin, vendor) {
+    const vendorURL = new URL(assertVendor(vendor));
```

Comment:
`URL` is not available everywhere. Use `parseUrl` from `url.js`.

Reply:
Ok, will do

Reply:
Done